### PR TITLE
[SITE] API Authentication Routes

### PR DESCRIPTION
**Pull request by benwerner01**
This PR sets up  the Next API with the relevant middleware and handlers to implement the following authentication related API routes:
 - `POST /api/sendLoginCode`: sends a login code to an existing BP user's email
 - `POST /api/loginWithLoginCode`: logs in a user using a provide login code
 - `GET /api/me`: retrieves the user object of the currently logged in user
 - `POST /api/logout`: logs out the currently authenticated user

Future Tasks:
- [setup AWS SES to send email invitations](https://app.asana.com/0/0/1201618515575266/f)

### CODE REVIEW

File: site/src/pages/api/sendLoginCode.api.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          msg: "Could not find user with the provided email",
+          value: email,
+        }),
+      );
+    }
+
+    if (await user.hasExceededLoginCodeRateLimit(db)) {
+      return res.status(403).json(
+        formatErrors({
+          msg: "You have exceeded the login code rate limit",
```
Comment (CiaranMn) [inline site/src/pages/api/sendLoginCode.api.ts:44]: - please wait xx minutes and try again

**Review by CiaranMn — APPROVED** [review]
Thanks Ben, looks good - a few minor comments added.

File: site/src/lib/model/verificationCode.model.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    return new DBRef(VerificationCode.COLLECTION_NAME, new ObjectId(this.id));
+  }
+
+  async incrementAttempts(db: Db) {
+    const { value } = await db
+      .collection<VerificationCodeDocument>(VerificationCode.COLLECTION_NAME)
+      .findOneAndUpdate(
+        { _id: new ObjectId(this.id) },
+        { $inc: { numberOfAttempts: 1 } },
+      );
```
[the remark this answers is not in the document]
Comment (benwerner01) [reply site/src/lib/model/verificationCode.model.ts:104]: Thanks for spotting - I'm still a bit unfamiliar with mongo

**Comment by CiaranMn** [comment]
One additional thing - we don't capture  a user's `shortname` anywhere here - I think it's fine to add in the signup flow, but just flagging so it doesn't get lost.

File: site/src/lib/model/verificationCode.model.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (benwerner01) [reply site/src/lib/model/verificationCode.model.ts:104]: FYI for the Node JS mongo package the option is `returnDocument: "after"`

**Comment by benwerner01** [comment]
> One additional thing - we don't capture a user's `shortname` anywhere here - I think it's fine to add in the signup flow, but just flagging so it doesn't get lost.

Thanks for flagging, I was going to include that in the API signup PR

File: site/README.md [hunk 3]
```diff
@@ -7,6 +7,46 @@ The public-facing [blockprotocol.org](https://blockprotocol.org) website serves
...
+- Request Response:
+  - `user`: the user currently authenticated with the API
+
+### `POST /api/logout` [authenticated]
+
+Logs out the currently authenticated user.
+
+- Request Response: `SUCCESS`
+
 ## Local development
```
[the remark this answers is not in the document]
Comment (benwerner01) [reply site/README.md:51]: Have done so 👍

**Pull request merged**